Web apps datepicker: upgrade and add to case search

orangejenny opened the pull request:
## Product Description
https://dimagi-dev.atlassian.net/browse/USH-1531 and https://dimagi-dev.atlassian.net/browse/USH-2009

This upgrades the datepicker widget in web apps to match HQ. The old widget doesn't support AMD. [This PR](https://github.com/dimagi/commcare-hq/pull/29310) introduced the new datepicker. The widget handles both dates and times.


![Screen Shot 2022-05-24 at 10 18 35 AM](https://user-images.githubusercontent.com/1486591/170060201-08e9042c-4132-4fce-ab52-78d0e3898518.png)
![Screen Shot 2022-05-24 at 10 18 44 AM](https://user-images.githubusercontent.com/1486591/170060203-8f1c4355-c539-4063-b5a3-f861c377af13.png)


This also adds a single date to the case search format options, because it's awkward UX to have to input a single date "range" for fields like DOB.

![Screen Shot 2022-05-24 at 10 18 59 AM](https://user-images.githubusercontent.com/1486591/170060277-ee0e9795-0d1e-419a-b545-0109ad1828d7.png)

I took a stab at implementing datetime questions in web apps and ran into formplayer issues, I think related to time zones, and reverted those changes.

depends on https://github.com/dimagi/commcare-core/pull/1115

## Feature Flag
The widget upgrade will affect all users.

The date format option in case search is behind the USH_CASE_CLAIM_UPDATES flag.

## Safety Assurance

### Safety story
Did some local testing, sending to QA.

### Automated test coverage

There's a basic suite XML test in this PR. The UI doesn't have tests.

### QA Plan

https://dimagi-dev.atlassian.net/browse/QA-4203

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

orangejenny committed:
Replaced old datepicker with new datepicker

orangejenny committed:
Removed old widget

orangejenny committed:
Removed most custom widget styles

orangejenny committed:
Added date widget support to web apps case search screen

orangejenny committed:
Added date widget in case search configuration support to app manager

orangejenny committed:
Added suite xml test coverage

orangejenny committed:
Updated other test that was affected by previous commit

orangejenny committed:
Updated yarn.lock

gherceg commented on corehq/apps/cloudcare/static/cloudcare/js/form_entry/entries.js:
```diff
@@ -762,61 +751,24 @@ hqDefine("cloudcare/js/form_entry/entries", function () {
-                timepicker: self.timepicker,
-                datepicker: self.datepicker,
+            self.$picker.datetimepicker({
+                date: self.answer() ? self.convertServerToClientFormat(self.answer()) : self.answer(),
```
nit: I know this is existing code, and maybe this is a weird JS thing, but is it possible to specify a value for the else case here? It's a bit strange to see `self.answer() ? self.convertServerToClientFormat(self.answer()) : self.answer()`

snopoke replied:
do you mean instead of just using `self.answer()` I think that would be `null`, `""` or `undefined`

gherceg replied:
Yeah so explicitly setting the date to one of those values if `self.answer()` evaluates to false. 
```suggestion
                date: self.answer() ? self.convertServerToClientFormat(self.answer()) : "",
```

orangejenny replied:
I hear you about this being weird, but I'm loath to change it without being certain what the "correct" falsy value to use is.

gherceg replied:
Yeah my only thought is would it be easy to test using each of the falsy values locally and see which, if any break this? If that isn't straightforward don't worry about.

orangejenny replied:
yeah, that sounds not terrible, I'll give it a shot

orangejenny replied:
474e723c88b

gherceg commented on corehq/apps/app_manager/templates/app_manager/partials/modules/case_search_property.html:
```diff
@@ -70,6 +70,10 @@ <h4 class="modal-title">
                     <div class="col-xs-12 col-sm-9">
                       <select class="form-control" data-bind="value: appearance">
                         <option value="">{% trans "Text" %}</option>
+                        {% if request|toggle_enabled:'USH_CASE_CLAIM_UPDATES' %}
+                          {# flag this one because it only works on web apps #}
```
Forgive my case search ignorance, but I'm having trouble interpreting this. Is it that setting a specific date for this case search config only works in web apps? And to ensure it is only available to projects with web apps enabled, we rely on this `USH_CASE_CLAIM_UPDATES` flag? Is there not a better way to make this available to any project that has web apps enabled?

orangejenny replied:
Yes, setting a specific date only works on web apps, since I haven't updated the android UI code. We don't have a good automated way to distinguish between mobile projects and web apps projects, this flag will just limit it to USH projects. It's possible I should just remove the comment.

gherceg replied:
Ah right! Makes sense. I like having the comment there, but might suggest:
```suggestion
                        {% if request|toggle_enabled:'USH_CASE_CLAIM_UPDATES' %}
                          {# limit to USH since datepicker is only available on web apps #}
```
Definitely a nit though, so feel free to ignore.

gherceg reviewed (commented):
Will let others more familiar approve 👍🏻

orangejenny committed:
Added missing template tags library

orangejenny committed:
Update corehq/apps/app_manager/templates/app_manager/partials/modules/case_search_property.html

Co-authored-by: Graham Herceg <gherceg@dimagi.com>

orangejenny committed:
Added widget event to query view

orangejenny committed:
Updated dp.change callback in query view

orangejenny committed:
Fixed lint error

orangejenny committed:
Updated weird logic to use NO_ANSWER

orangejenny committed:
Updated clearAction to handle date widget

orangejenny committed:
Fixed QA-4227

Both change and dp.change events fire, and running the change handler twice was a problem.
This change also allows removing the earlier fix that added a call to _.delay.

orangejenny committed:
Added clear button

orangejenny committed:
Merge branch 'master' into jls/upgrade-web-apps-datepicker

orangejenny committed:
Updated display options

orangejenny committed:
Updated app preview styles and added close button
